File: src/renderer/tui/style.rs
use ratatui::style::{Color, Modifier, Style};

pub fn parse_style(input: &str) -> Style {
    let mut style = Style::default();
    for part in input.split(',') {
        match part.trim().to_lowercase().as_str() {
            "bold" => style = style.add_modifier(Modifier::BOLD),
            "italic" => style = style.add_modifier(Modifier::ITALIC),
            "underline" => style = style.add_modifier(Modifier::UNDERLINED),
            "faint" | "dim" => style = style.add_modifier(Modifier::DIM),
            _ => {}
        }
    }
    style
}

pub fn parse_color(input: &str) -> Option<Color> {
    let input = input.trim();
    if input.is_empty() {
        return None;
    }
    input.parse().ok()
}

pub fn build_style(style_str: &str, color: Option<&str>) -> Style {
    let mut style = parse_style(style_str);
    if let Some(color) = color.and_then(parse_color) {
        style = style.fg(color);
    }
    style
}

#[cfg(test)]
mod tests {
    use super::*;

    // --- parse_style tests ---

    #[test]
    fn test_parse_style_bold() {
        assert_eq!(parse_style("bold"), Style::default().add_modifier(Modifier::BOLD));
    }

    #[test]
    fn test_parse_style_faint() {
        assert_eq!(parse_style("faint"), Style::default().add_modifier(Modifier::DIM));
    }

    #[test]
    fn test_parse_style_combined() {
        assert_eq!(
            parse_style("bold,italic"),
            Style::default().add_modifier(Modifier::BOLD | Modifier::ITALIC)
        );
    }

    #[test]
    fn test_parse_style_all_modifiers() {
        let style = parse_style("bold,italic,underline,faint");
        let expected = Style::default()
            .add_modifier(Modifier::BOLD | Modifier::ITALIC | Modifier::UNDERLINED | Modifier::DIM);
        assert_eq!(style, expected);
    }

    #[test]
    fn test_parse_style_with_spaces() {
        assert_eq!(
            parse_style(" bold , italic "),
            Style::default().add_modifier(Modifier::BOLD | Modifier::ITALIC)
        );
    }

    #[test]
    fn test_parse_style_unknown_ignored() {
        assert_eq!(parse_style("bold,unknown"), Style::default().add_modifier(Modifier::BOLD));
    }

    // --- parse_color tests ---

    #[test]
    fn test_parse_color_hex() {
        assert_eq!(parse_color("#ff5500"), Some(Color::Rgb(255, 85, 0)));
    }

    #[test]
    fn test_parse_color_hex_black() {
        assert_eq!(parse_color("#000000"), Some(Color::Rgb(0, 0, 0)));
    }

    #[test]
    fn test_parse_color_ansi_index() {
        assert_eq!(parse_color("245"), Some(Color::Indexed(245)));
    }

    #[test]
    fn test_parse_color_named_red() {
        assert_eq!(parse_color("red"), Some(Color::Red));
    }

    #[test]
    fn test_parse_color_named_gray() {
        assert_eq!(parse_color("gray"), Some(Color::Gray));
    }

    #[test]
    fn test_parse_color_named_white() {
        assert_eq!(parse_color("white"), Some(Color::White));
    }

    #[test]
    fn test_parse_color_empty() {
        assert_eq!(parse_color(""), None);
    }

    #[test]
    fn test_parse_color_invalid() {
        assert_eq!(parse_color("notacolor"), None);
    }

    // --- build_style tests ---

    #[test]
    fn test_build_style_with_color() {
        let style = build_style("bold", Some("red"));
        assert_eq!(style, Style::default().add_modifier(Modifier::BOLD).fg(Color::Red));
    }

    #[test]
    fn test_build_style_without_color() {
        let style = build_style("faint", None);
        assert_eq!(style, Style::default().add_modifier(Modifier::DIM));
    }
}

File: src/renderer/tui/style.ts
// Replicates `ratatui` Style/Color as plain data so tests use structural equality.

export type NamedColor =
  | "reset"
  | "black"
  | "red"
  | "green"
  | "yellow"
  | "blue"
  | "magenta"
  | "cyan"
  | "gray"
  | "darkgray"
  | "lightred"
  | "lightgreen"
  | "lightyellow"
  | "lightblue"
  | "lightmagenta"
  | "lightcyan"
  | "white";

export type Color =
  | { type: "named"; name: NamedColor }
  | { type: "rgb"; r: number; g: number; b: number }
  | { type: "indexed"; index: number };

export interface Style {
  bold: boolean;
  italic: boolean;
  underline: boolean;
  dim: boolean;
  fg: Color | null;
}

export const RESET = "\x1b[0m";

export function defaultStyle(): Style {
  return { bold: false, italic: false, underline: false, dim: false, fg: null };
}

/**
 * Mirrors `parse_style`: split on ',', trim + lowercase each part. `faint`/`dim`
 * both map to the `dim` modifier. Unknown tokens are ignored.
 */
export function parseStyle(input: string): Style {
  const style = defaultStyle();
  for (const part of input.split(",")) {
    switch (part.trim().toLowerCase()) {
      case "bold":
        style.bold = true;
        break;
      case "italic":
        style.italic = true;
        break;
      case "underline":
        style.underline = true;
        break;
      case "faint":
      case "dim":
        style.dim = true;
        break;
      default:
        break;
    }
  }
  return style;
}

const NAMED_COLORS: ReadonlySet<NamedColor> = new Set<NamedColor>([
  "reset",
  "black",
  "red",
  "green",
  "yellow",
  "blue",
  "magenta",
  "cyan",
  "gray",
  "darkgray",
  "lightred",
  "lightgreen",
  "lightyellow",
  "lightblue",
  "lightmagenta",
  "lightcyan",
  "white",
]);

/**
 * Replicates Rust `u8::from_str`: an optional leading `+`, then ASCII digits
 * only, with the value in 0..=255. Returns null for anything else (whitespace,
 * leading `-`, non-digits, out of range, empty).
 */
function parseU8(s: string): number | null {
  let body = s;
  if (body.startsWith("+")) {
    body = body.slice(1);
  }
  if (body.length === 0) {
    return null;
  }
  for (const ch of body) {
    if (ch < "0" || ch > "9") {
      return null;
    }
  }
  const value = Number(body);
  if (value > 255) {
    return null;
  }
  return value;
}

/** Replicates `ratatui` `parse_hex_color`: requires `#` prefix and length 7. */
function parseHexColor(input: string): { r: number; g: number; b: number } | null {
  if (!input.startsWith("#") || input.length !== 7) {
    return null;
  }
  const parseByte = (hex: string): number | null => {
    if (!/^[0-9a-fA-F]{2}$/.test(hex)) {
      return null;
    }
    return parseInt(hex, 16);
  };
  const r = parseByte(input.slice(1, 3));
  const g = parseByte(input.slice(3, 5));
  const b = parseByte(input.slice(5, 7));
  if (r === null || g === null || b === null) {
    return null;
  }
  return { r, g, b };
}

/**
 * Mirrors `parse_color` + ratatui `Color::from_str` (case-insensitive, trimmed):
 * - empty → null
 * - normalized named color (incl. grey/silver/bright aliases) → named
 * - all-digit 0..=255 (on the raw trimmed input) → indexed
 * - `#rrggbb` → rgb
 * - otherwise → null
 */
export function parseColor(input: string): Color | null {
  const trimmed = input.trim();
  if (trimmed.length === 0) {
    return null;
  }

  // Named-color match operates on the normalized string (ratatui Color::from_str).
  const normalized = trimmed
    .toLowerCase()
    .replace(/[ \-_]/g, "")
    .replace(/bright/g, "light")
    .replace(/grey/g, "gray")
    .replace(/silver/g, "gray")
    .replace(/lightblack/g, "darkgray")
    .replace(/lightwhite/g, "white")
    .replace(/lightgray/g, "white");

  if (NAMED_COLORS.has(normalized as NamedColor)) {
    return { type: "named", name: normalized as NamedColor };
  }

  // Integer + hex branches use the raw (trimmed) input, matching Rust.
  const index = parseU8(trimmed);
  if (index !== null) {
    return { type: "indexed", index };
  }

  const rgb = parseHexColor(trimmed);
  if (rgb !== null) {
    return { type: "rgb", r: rgb.r, g: rgb.g, b: rgb.b };
  }

  return null;
}

/** Mirrors `build_style`: parse modifiers, then apply fg if a color parses. */
export function buildStyle(styleStr: string, color: string | null | undefined): Style {
  const style = parseStyle(styleStr);
  if (color !== null && color !== undefined) {
    const parsed = parseColor(color);
    if (parsed !== null) {
      style.fg = parsed;
    }
  }
  return style;
}

const NAMED_ANSI_FG: Record<NamedColor, string> = {
  reset: "39",
  black: "30",
  red: "31",
  green: "32",
  yellow: "33",
  blue: "34",
  magenta: "35",
  cyan: "36",
  gray: "37",
  darkgray: "90",
  lightred: "91",
  lightgreen: "92",
  lightyellow: "93",
  lightblue: "94",
  lightmagenta: "95",
  lightcyan: "96",
  white: "97",
};

/**
 * Builds the opening SGR escape sequence for a style. Returns "" when the style
 * carries no attributes and no foreground color. Named foreground codes are
 * best-effort and not asserted by tests.
 */
export function styleToAnsi(style: Style): string {
  const codes: string[] = [];
  if (style.bold) {
    codes.push("1");
  }
  if (style.dim) {
    codes.push("2");
  }
  if (style.italic) {
    codes.push("3");
  }
  if (style.underline) {
    codes.push("4");
  }
  if (style.fg !== null) {
    switch (style.fg.type) {
      case "rgb":
        codes.push(`38;2;${style.fg.r};${style.fg.g};${style.fg.b}`);
        break;
      case "indexed":
        codes.push(`38;5;${style.fg.index}`);
        break;
      case "named":
        codes.push(NAMED_ANSI_FG[style.fg.name]);
        break;
      default:
        break;
    }
  }
  if (codes.length === 0) {
    return "";
  }
  return `\x1b[${codes.join(";")}m`;
}
